File: Prak 2019/Minggu 1/Time.cpp
#include "Time.hpp"

int Time::n_time = 0;

Time::Time(){
    this->hour = 0;
    this->minute = 0;
    this->second = 0;
    n_time++;
}

Time::Time(int h, int m, int s){
    this->hour = h;
    this->minute = m;
    this->second = s;
    n_time++;
}

Time::Time(const Time& Time){
    this->hour = Time.hour;
    this->minute = Time.minute;
    this->second = Time.second;
    n_time++;
}

void Time::setHour(int h){
    this->hour = h;
}

void Time::setMinute(int m){
    this->minute = m;
}

void Time::setSecond(int s){
    this->minute = s;
}

int Time::getHour() const{
    return this->hour;
}

int Time::getMinute() const{
    return this->minute;
}

int Time::getSecond() const{
    return this->second;
}

int Time::countTimeInstance(){
    return n_time;
}

int Time::convertToSecond() const{
    return this->hour * 3600 + this->minute * 60 + this->second;
}

void Time::printTime(){
    if (this->hour < 10) {
        cout << "0" << this->hour << ":";
    } else {
        cout << this->hour << ":";
    }
    if (this->minute < 10){
        cout << "0" << this->minute << ":";
    } else {
        cout << this->minute << ":";
    }
    if (this->second < 10){
        cout << "0" << this->second << endl;;
    } else {
        cout << this->second << endl;
    }
}

void Time::addTime(const Time& t){
    int newTime = this->convertToSecond() + t.convertToSecond();
    int newHour = (newTime / 3600) % 24;
    newTime -= newHour * 3600;
    int newMinute = (newTime / 60) % 60;
    newTime -= newMinute * 60;
    int newSeconds = newMinute % 60;
    this->setHour(newHour);
    this->setMinute(newMinute);
    this->setSecond(newSeconds);
}

void Time::decreaseTime(const Time& t){
    if (this->laterThan(t)){
        int newTime = this->convertToSecond() - t.convertToSecond();
        int newHour = (newTime / 3600) % 24;
        newTime -= newHour * 3600;
        int newMinute = (newTime / 60) % 60;
        newTime -= newMinute * 60;
        int newSeconds = newMinute % 60;
        this->setHour(newHour);
        this->setMinute(newMinute);
        this->setSecond(newSeconds);
    }
}

bool Time::earlierThan(const Time& t){
    return this->convertToSecond() < t.convertToSecond();
}

bool Time::laterThan(const Time &t){
    return this->convertToSecond() > t.convertToSecond();
}

File: Praktikum-2021/Minggu 7 - Tutorial 4/1/Time.java
public class Time {
    private int hour;
    private int minute;
    private int second;

    public Time() {
        // Set seluruh atribut dengan nilai 0
        this.hour = 0;
        this.minute = 0;
        this.second = 0;
    }

    public Time(int hour, int minute, int second) {
        this.hour = hour;
        this.minute = minute;
        this.second = second;
    }

    public Time(Time t) {
        this.hour = t.hour;
        this.minute = t.minute;
        this.second = t.second;
        // Set atribut dengan atribut milik t
    }

    public int getHour() {
        return this.hour;
    }

    public void setHour(int hour) {
        this.hour = hour;
    }

    public int getMinute() {
        return this.minute;
    }

    public void setMinute(int minute) {
        this.minute = minute;
    }

    public int getSecond() {
        return this.second;
    }

    public void setSecond(int second) {
        this.second = second;
    }

    public int convertToSecond() {
        return this.hour * 3600 + this.minute * 60 + this.second;
    }

    public Time add(Time t) {
        // Return penjumlahan dua objek jam, yaitu *this* dan t.
        // Apabila hasil penjumlahan melebihi 23:59:59, maka Anda harus mengonversinya seperti tampilan jam digital asli.
        // Contoh : 24:00:00 dituliskan 00:00:00, 25:00:00 dituliskan 01:00:00, dst.
        int seconds = (t.convertToSecond() + this.convertToSecond()) % 86400;
        int hour = seconds / 3600;
        int minutes = (seconds - hour * 3600) / 60;
        int second = seconds - hour * 3600 - minutes * 60;
        return new Time(hour, minutes, second);
    }

    public Time minus(Time t) {
        // Return selisih antara dua objek jam, yaitu *this* dan t.
        // Perhitungan selisih hanya dapat dilakukan jika objek ruas kiri lebih akhir dari atau sama dengan objek ruas kanan.
        // Jika objek ruas kiri lebih awal, maka kembalikan nilai objek ruas kiri.
        // Contoh: 00:00:01 - 00:00:02 = 00:00:01
        if (this.lessThan(t)){
            return this;
        } else {
            int seconds = this.convertToSecond() - t.convertToSecond();
            int hour = seconds / 3600;
            int minutes = (seconds - hour * 3600) / 60;
            int second = seconds - hour * 3600 - minutes * 60;
            return new Time(hour, minutes, second);
        }
    }

    public boolean lessThan(Time t) {
        // Returns true jika *this* < t menurut definisi waktu
        return this.convertToSecond() < t.convertToSecond();
    }

    public boolean greaterThan(Time t) {
        // Returns true jika *this* > t menurut definisi waktu
        return this.convertToSecond() > t.convertToSecond();
    }

    public void printTime() {
        // Menuliskan jam dalam format hh:mm:ss diakhiri dengan newline. Perhatikan bahwa setiap parameter harus dituliskan 2 digit.
        System.out.println(String.format("%02d", this.hour) + ":" + String.format("%02d", this.minute) + ":" + String.format("%02d", this.second));
    }
}